### PR TITLE
Concise and static Python imports

### PR DESCRIPTION
**Pull request by Helveg**
Removed all forms of dynamic module importing and `from X import *`. If the need for dynamic modules arises again they can be defined as a lazy loaded module such as `spatial` with some extra indications for the user when missing optional modules are accessed:

```
>>> nest.optional.feature()
ImportError: This functionality requires the optional `optional` module. See <nest-docs> to install it.
```

### PR summary
* Replaced `ast`/`compile` based dynamic importing by `importlib`.
  * `ast`/`compile` sidesteps import machinery.
  * Added mechanism for _optional_ modules `_lazy_module_property("feature", optional=True, optional_hint="Please install feature.")`
* Removed `hl_api`
* Replaced `from X import *` calls with explicit imports.
* Simplified test imports/discovery, removed gratuitous test helper file.

### CODE REVIEW

**Comment by Helveg** [comment]
@jougs I hope you don't mind I went ahead and removed `hl_api.py` prematurely, the code in it would only confuse developers. I checked the codebase and it wasn't used anywhere except for a couple of times as `nest.hl_api.serializable`, I think also #2157 changes those to just `nest.serializable`. 

The change could have affected users that do funky things with deep imports, but I just added `nest` as `nest.hl_api` so they can safely continue importing from there until we clean that up in 4.0. I'm actually all for removing it altogether because I doubt many people actually use these deep imports.

**Comment by jougs** [comment]
No worries! I'm always in favor of cleaning things up and not adding backwards-compatibility-cruft.

As this PR won't make it into 3.1, I suggest, you create another small PR, adding a sentence to the [release notes](https://github.com/nest/nest-simulator/blob/master/doc/userdoc/release_notes/v3.1/index.rst) that the `hl_api` namespace will go away with 3.2.

**Comment by Helveg** [comment]
(side note on #2157 to @jougs and in general also for @heplesser and others)

All of our problems with the reference to the old nest module seem to be disguised circular import errors that due to a bug in Python makes it pick up the unreplaced object that it compiled and executed in order to even get to the point where we try to replace it but then because of circular imports just silently doesn't finish.

I think the codebase should be thoroughly vetted for code that gets executed at import time and also the nest submodules should **not** import nest itself, but their sibling modules.

File: pynest/nest/__init__.py [hunk 1]
```diff
@@ -406,14 +427,22 @@ def __dir__(self):
...
-# We manipulate the nest module instance through its `__dict__` (= vars())
+# Manipulate the nest module instance through its `__dict__` (= vars())
 _module_dict = vars(_module)
 # Copy over the original module attributes to preverse all interpreter given
 # magic attributes such as `__name__`, `__path__`, `__package__`, ...
 _module_dict.update(_original_module_attrs)
 
-# Import public API of `.hl_api` into the nest module instance
-_rel_import_star(_module_dict, ".hl_api")
+# Import public APIs of submodules into the `nest.` namespace
```
[the remark this answers is not in the document]
Comment (Helveg) [reply pynest/nest/__init__.py]: @heplesser The diff is a bit messy but what I did was execute all the module initialization to `__init__` as you proposed, and that allowed me to shuffle some other parts around too (such as the `_rel_import_star` and `_lazy_module_propery` defs).

**Review by heplesser — APPROVED** [review]

**Pull request merged**